File: rover.rs
	pub enum Direction {
		N,
		S,
		E,
		W
	}

	struct Point {
		pub x: i32,
		pub y: i32
	}

	pub struct Rover {
		pub position: Point,
		direction: Direction
	}

	impl Rover {
		pub fn new(x: i32, y: i32, d: Direction) -> Rover {
			println!("     Constructing a rover... ({0},{1}) {2}", x, y, d as i32);
			Rover {position: Point{x: x,y: y}, direction: d}
		}

		fn move(&mut self) -> () {
			match self.direction {
				N => {
					self.position.y += 1;
					println!("     Moving North...");
				},
				S => {
					self.position.y -= 1;
					println!("     Moving South...");
				},
				E => {
					self.position.x += 1;
					println!("     Moving East...");
				},
				W => {
					self.position.x -= 1;
					println!("     Moving West...");
				}
			}
			println!("     new robot position: ({0},{1}) {2}", self.position.x, self.position.y, self.direction as i32);
		}

		fn turn_left(&mut self) -> () {
			match self.direction {
				N => self.direction = W,
				S => self.direction = E,
				E => self.direction = N,
				W => self.direction = S
			}
		}

		fn turn_right(&mut self) -> () {
			match self.direction {
				N => self.direction = E,
				S => self.direction = W,
				E => self.direction = S,
				W => self.direction = N
			}
		}

		pub fn travel(&mut self, moves: &str) -> bool {
			let mut success = true;

			for c in moves.chars() {
				match c {
					'L' => self.turn_left(),
					'R' => self.turn_right(),
					'M' => self.move(),
					_ => {
						println!("Unknown command: {}", c);
						success = false;
					}
				}
			}

			success
		}

		pub fn print(&self) -> () {
			println!("     Robot position: ({0},{1}) {2}", self.position.x, self.position.y, self.direction as i32);
		}
	}

File: rover.rb
class Point
	attr_accessor :x,:y

	def initialize(x,y)
		@x = x
		@y = y
	end
end

class Rover
	attr_accessor :loc, :heading

	def initialize(location, heading=:N)
		@loc = location
		@heading = heading
	end

	def turn_left
		case @heading
		when :N
			@heading = :E
		when :E
			@heading = :S
		when :S
			@heading = :W
		when :W
			@heading = :N
		end
	end

	def turn_right
		case @heading
		when :N
			@heading = :W
		when :W
			@heading = :S
		when :S
			@heading = :E
		when :E
			@heading = :N
		end
	end

	def forward
		case @heading
		when :N
			@loc.y += 1
		when :W
			@loc.x += 1
		when :S
			@loc.y -= 1
		when :E
			@loc.x -= 1
		end
	end

	def to_s
		super.to_s
		puts "Rover Position = (#{@loc.x}, #{@loc.y}) heading #{@heading}"
	end
end

class Mars
	def initialize(width, height)
		@width = width
		@height = height
		@rovers = []
	end

	def add_rover(initial_pos, heading)
		@rovers << Rover.new(initial_pos, heading)
		return @rovers[@rovers.length-1]
	end

	def move_rover(rover, cmds)
		# keep move code in Mars class as it can detect collisions with other rovers
		cmds.split("").each do |c|
			case c
			when 'L'
				rover.turn_left
			when 'R'
				rover.turn_right
			when 'M'
				rover.forward
				if rover.loc.x < 0 or rover.loc.x > @width or rover.loc.y < 0 or rover.loc.y > @height
					raise "Rover Crash!!!"
				end
			else
				raise "Unknown Command to Rover: #{c}"
			end
		end
	end

	def print_rovers
		for rover in @rovers
			puts rover
		end
	end
end

mars = Mars.new(5, 5)
rover = mars.add_rover(Point.new(1,2), :N)
mars.move_rover(rover, "LMLMLMLMM")
rover = mars.add_rover(Point.new(3,3), :E)
mars.move_rover(rover, "MMRMMRMRRM")
rover = mars.add_rover(Point.new(0,0), :S)
mars.move_rover(rover, "MMMMM")
mars.print_rovers
